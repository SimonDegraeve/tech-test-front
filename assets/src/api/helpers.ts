import { candidatePositionInterval } from './config'
import { Candidate } from './types'

// Calculate the new candidate position based on ascending or descending order
export const getCandidatePosition = (
  candidates: Candidate[],
  index: number,
  isDescendingOrder: boolean
) => {
  const previousCandidatePosition = candidates[index - 1]?.position ?? 0
  const defaultNextCandidatePosition = (index || 1) * candidatePositionInterval
  let nextCandidatePosition
  let newCandidatePosition

  // If the order is descending, the new candidate position should be between the previous and next candidate
  if (isDescendingOrder) {
    nextCandidatePosition = candidates[index]?.position ?? defaultNextCandidatePosition

    // The new position should be at least one position after the previous candidate
    // and at most half-way before the next candidate
    newCandidatePosition = Math.max(
      previousCandidatePosition + 1,
      nextCandidatePosition - Math.floor((nextCandidatePosition - previousCandidatePosition) / 2)
    )
  }
  // If the order is ascending, the new candidate position should be after the next candidate
  else {
    nextCandidatePosition = candidates[index + 1]?.position ?? defaultNextCandidatePosition

    // The new position should be at least half-way after the next candidate
    newCandidatePosition =
      nextCandidatePosition + Math.floor((previousCandidatePosition - nextCandidatePosition) / 2)
  }

  return newCandidatePosition
}
