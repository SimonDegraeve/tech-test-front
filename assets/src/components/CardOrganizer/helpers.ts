import { UniqueIdentifier } from '@dnd-kit/core'

// Type matching non exported `AnyData` from `@dnd-kit/core`, `any` is necessary to match the type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ItemData = { id: UniqueIdentifier; [key: string]: any }

export const emptyItemPlaceholderIdPrefix = '__empty__'

export type EmptyItemPlaceholder<Column extends string> = {
  id: `${typeof emptyItemPlaceholderIdPrefix}-${Column}`
}

export function isEmptyItemPlaceholder<Column extends string>(
  item: ItemData | EmptyItemPlaceholder<Column>
): item is EmptyItemPlaceholder<Column> {
  return item.id.toString().startsWith(emptyItemPlaceholderIdPrefix)
}
