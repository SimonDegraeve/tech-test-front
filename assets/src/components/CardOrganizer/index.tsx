import { forwardRef, memo, ReactNode, useEffect, useRef, useState } from 'react'
import { Text } from '@welcome-ui/text'
import { Flex } from '@welcome-ui/flex'
import { Badge } from '@welcome-ui/badge'
import { CSS } from '@dnd-kit/utilities'
import {
  closestCenter,
  DndContext,
  DragOverlay,
  KeyboardSensor,
  PointerSensor,
  useSensor,
  useSensors,
  DragEndEvent,
  DragStartEvent,
  UniqueIdentifier,
} from '@dnd-kit/core'
import {
  arrayMove,
  SortableContext,
  sortableKeyboardCoordinates,
  useSortable,
  verticalListSortingStrategy,
} from '@dnd-kit/sortable'
import {
  EmptyItemPlaceholder,
  emptyItemPlaceholderIdPrefix,
  isEmptyItemPlaceholder,
} from './helpers'

export type CardOrganizerProps<Column extends string, Item extends ItemData> = {
  columns: Column[]
  items: ItemRegistry<Column, Item>
  renderCard: (item: Item) => ReactNode
  renderColumnLabel?: (column: Column, count: number) => ReactNode
  onChange: (columnUpdate: ColumnUpdate<Column, Item>) => Promise<void>
}

function CardOrganizer<Column extends string, Item extends ItemData>({
  columns,
  items: parentItems,
  renderCard,
  renderColumnLabel,
  onChange,
}: CardOrganizerProps<Column, Item>) {
  const [internalActiveItem, setInternalActiveItem] = useState<
    ItemWithSortableData<Column, Item> | undefined
  >()

  /**
   * `internalItems` is used to manage updates when dragging between columns
   * it is instantiated with the `parentItems` when the drag starts
   * and it is reset when the drag ends if the component is still mounted
   */
  const [internalItems, setInternalItems] = useState<ItemRegistry<Column, Item> | undefined>()

  const isMounted = useRef(false)
  useEffect(() => {
    isMounted.current = true
    return () => {
      isMounted.current = false
    }
  }, [])

  const sensors = useSensors(
    useSensor(PointerSensor),
    // Keyboard navigation for improved accessibility
    useSensor(KeyboardSensor, { coordinateGetter: sortableKeyboardCoordinates })
  )

  function handleDragStart(event: DragStartEvent) {
    const { active } = event
    const activeItem = toCardOrganizerItem<Column, Item>(active.data.current)
    setInternalActiveItem(activeItem)
    setInternalItems(parentItems)
  }

  function handleDragEnd(event: DragEndEvent) {
    const { active, over } = event
    const activeItem = toCardOrganizerItem<Column, Item>(active.data.current)
    const overItem = toCardOrganizerItem<Column, Item>(over?.data.current)

    if (activeItem && overItem) {
      // Notify the parent component of the change
      onChange({
        from: {
          ...activeItem,
          // Remove empty item placeholder from the items list
          sortable: {
            ...activeItem.sortable,
            items: activeItem.sortable.items.filter(id => !isEmptyItemPlaceholder({ id })),
          },
        },
        to: {
          ...overItem,
          // Remove empty item placeholder from the items list
          sortable: {
            ...overItem.sortable,
            items: overItem.sortable.items.filter(id => !isEmptyItemPlaceholder({ id })),
          },
        },
      }).finally(() => {
        // Note: Could use `requestAnimationFrame` instead of `setTimeout` to wait for the next frame
        setTimeout(() => {
          if (
            isMounted.current &&
            (!internalActiveItem || internalActiveItem.id === activeItem.id)
          ) {
            setInternalItems(undefined)
          }
        }, 16 /* 1 frame */)
      })

      // Move the item in internal items to reflect the change
      if (activeItem.sortable.containerId === overItem.sortable.containerId) {
        setInternalItems(items => {
          const oldIndex = activeItem.sortable.index
          const newIndex = overItem.sortable.index
          const column = activeItem.sortable.containerId
          return {
            ...items,
            [column]: arrayMove(items?.[column] ?? [], oldIndex, newIndex),
          }
        })
      }
    } else {
      setInternalItems(undefined)
    }

    setInternalActiveItem(undefined)
  }

  function onDragOver(event: DragEndEvent) {
    const { active, over } = event
    const activeItem = toCardOrganizerItem<Column, Item>(active.data.current)
    const overItem = toCardOrganizerItem<Column, Item>(over?.data.current)

    if (activeItem && overItem) {
      // If the item is dragged to a different column, update the internal items
      if (activeItem.sortable.containerId !== overItem.sortable.containerId) {
        setInternalItems(items => {
          const oldIndex = activeItem.sortable.index
          const newIndex = overItem.sortable.index
          const fromColumn = activeItem.sortable.containerId
          const toColumn = overItem.sortable.containerId
          const fromItems = items?.[fromColumn] ?? []
          const toItems = items?.[toColumn] ?? []
          const item = fromItems[oldIndex]

          // Remove the item from the original column..
          const newFromItems = [...fromItems.slice(0, oldIndex), ...fromItems.slice(oldIndex + 1)]
          // ...and insert it into the new column
          const newToItems = [...toItems.slice(0, newIndex), item, ...toItems.slice(newIndex)]

          return {
            ...items,
            [fromColumn]: newFromItems,
            [toColumn]: newToItems,
          }
        })
      }
    }
  }

  function onDragCancel() {
    setInternalActiveItem(undefined)
    setInternalItems(undefined)
  }

  return (
    <DndContext
      sensors={sensors}
      collisionDetection={closestCenter}
      onDragStart={handleDragStart}
      onDragEnd={handleDragEnd}
      onDragOver={onDragOver}
      onDragCancel={onDragCancel}
    >
      <Flex p={'20 0'} overflow="hidden" flexGrow={1}>
        <Flex gap={10} overflow="hidden">
          <DragOverlay>
            {internalActiveItem ? (
              <Item style={{ transform: 'rotate(3deg)' }}>{renderCard(internalActiveItem)}</Item>
            ) : null}
          </DragOverlay>

          {columns.map(column => {
            const items: (Item | EmptyItemPlaceholder<Column>)[] =
              internalItems?.[column] ?? parentItems[column] ?? []

            // Add an empty item placeholder if the column is empty,
            // this is necessary to allow dragging items to an empty column because we need somethinng to collide with
            if (!items.length) {
              items.push({ id: `${emptyItemPlaceholderIdPrefix}-${column}` })
            }

            return (
              <SortableContext
                id={column}
                key={column}
                items={items}
                strategy={verticalListSortingStrategy}
              >
                <MemoizedCardColumn
                  key={column}
                  column={column}
                  items={items}
                  activeItem={internalActiveItem}
                  renderCard={renderCard}
                  renderColumnLabel={renderColumnLabel}
                />
              </SortableContext>
            )
          })}
        </Flex>
      </Flex>
    </DndContext>
  )
}

export default CardOrganizer

/**
 * CardOrganizer types and helpers
 */

// Type matching non exported `AnyData` from `@dnd-kit/core`, `any` is necessary to match the type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ItemData = { id: UniqueIdentifier; [key: string]: any }

type ItemRegistry<Column extends string, Item extends ItemData> = Partial<Record<Column, Item[]>>

export type ColumnUpdate<Column extends string, Item extends ItemData> = {
  from: ItemWithSortableData<Column, Item>
  to: ItemWithSortableData<Column, Item>
}

type ItemWithSortableData<Column extends string, Item extends ItemData> = Item & {
  sortable: { containerId: Column; index: number; items: UniqueIdentifier[] }
}

function toCardOrganizerItem<Column extends string, Item extends ItemData>(
  item: { [key: string]: unknown } | undefined
) {
  // Transforms `AnyData` from dnd-kit/core into our own more specific type
  return item as ItemWithSortableData<Column, Item> | undefined
}

/**
 * CardColumn & MemoizedCardColumn components
 */

function CardColumn<Column extends string, Item extends ItemData>({
  column,
  items,
  activeItem,
  renderCard,
  renderColumnLabel,
}: {
  column: Column
  items: (Item | EmptyItemPlaceholder<Column>)[]
  activeItem: ItemWithSortableData<Column, Item> | undefined
  renderCard: (item: Item) => ReactNode
  renderColumnLabel?: (column: Column, count: number) => ReactNode
  maxColumnSize?: number
}) {
  const itemsCount = items.filter(item => !isEmptyItemPlaceholder(item)).length

  return (
    <Flex
      w={300}
      border={1}
      backgroundColor="white"
      borderColor="neutral-30"
      borderRadius="md"
      overflow="hidden"
      flexDirection="column"
      data-testid={`column-${column}`}
    >
      <Flex
        p={10}
        borderBottom={1}
        borderColor="neutral-30"
        alignItems="center"
        justify="space-between"
      >
        {renderColumnLabel?.(column, itemsCount) ?? (
          <Text color="black" m={0} textTransform="capitalize" fontWeight="600">
            {column}
          </Text>
        )}

        <Badge data-testid={`column-counter-${column}`}>{itemsCount}</Badge>
      </Flex>

      <Flex overflow="hidden" h="100%">
        <Flex direction="column" p={10} minHeight={100} gap={10} overflow="auto" w="100%">
          {items.map(item => (
            <SortableItem
              key={item.id}
              data={item}
              style={item.id === activeItem?.id ? { opacity: 0.5 } : {}}
            >
              {isEmptyItemPlaceholder(item) ? null : renderCard(item)}
            </SortableItem>
          ))}
        </Flex>
      </Flex>
    </Flex>
  )
}

const MemoizedCardColumn = memo(
  CardColumn,
  (
    { column: prevColumn, items: prevItems, activeItem: prevActiveItem },
    { column: nextColumn, items: nextItems, activeItem: nextActiveItem }
  ) => {
    const hasSameColumn = prevColumn === nextColumn

    // Warning: The column will not update if the data of the items changes
    const hasSameItems =
      prevItems.map(({ id }) => id).join('') === nextItems.map(({ id }) => id).join('')

    const hasSameActiveColumn =
      !!prevItems.find(({ id }) => id === prevActiveItem?.id) ===
      !!nextItems.find(({ id }) => id === nextActiveItem?.id)

    return hasSameColumn && hasSameItems && hasSameActiveColumn
  }
) as typeof CardColumn

/**
 * Item & SortableItem components
 */

export type ItemProps = React.PropsWithChildren<React.HTMLAttributes<HTMLDivElement>>

export const Item = forwardRef(
  ({ children, ...props }: ItemProps, ref: React.LegacyRef<HTMLDivElement> | undefined) => {
    return (
      <div {...props} ref={ref}>
        {children}
      </div>
    )
  }
)

export type SortableItemProps<DragItemData extends ItemData> = ItemProps & {
  data: DragItemData
}

export function SortableItem<DragItemData extends ItemData>({
  data,
  style: extraStyle,
  children,
}: SortableItemProps<DragItemData>) {
  const { attributes, listeners, setNodeRef, transform, transition } = useSortable({
    id: data.id,
    data,
  })

  const style = {
    transform: `${CSS.Transform.toString(transform) ?? ''} ${extraStyle?.transform ?? ''}`,
    transition,
    ...extraStyle,
  }

  return (
    <Item ref={setNodeRef} style={style} {...attributes} {...listeners}>
      {children}
    </Item>
  )
}
