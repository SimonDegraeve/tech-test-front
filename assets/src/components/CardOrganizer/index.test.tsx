import { describe, expect, test, vi } from 'vitest'
import { act, fireEvent } from '@testing-library/react'
import { UniqueIdentifier } from '@dnd-kit/core'

import { render } from '../../test-utils'
import CardOrganizer from '../../components/CardOrganizer'
import { isEmptyItemPlaceholder } from './helpers'

/**
 * Note: the tests rely on the keyboard navigation as the the DOM is emulated by the testing library (jsdom) we cannot trigger real drag and drop events.
 * In order to test real drag and drop events, we should use a real browser in e2e tests (with utils like Playwright or Cypress).
 * See: https://github.com/clauderic/dnd-kit/issues/261#issuecomment-844597229
 */
describe('CardOrganizer', () => {
  type Item = { id: UniqueIdentifier; name: string }

  const defaultProps = {
    columns: ['A', 'B', 'C'],
    items: {
      A: [
        { id: 1, name: 'Item 1' },
        { id: 2, name: 'Item 2' },
        { id: 3, name: 'Item 3' },
      ],
      B: [],
      C: [],
    },
    onChange: vi.fn().mockResolvedValue(new Promise(() => {})),
    renderCard: (item: Item) => item.name,
  }

  test('renders every columns with their counters', () => {
    const { getByText, getByTestId } = render(
      <CardOrganizer<(typeof defaultProps.columns)[number], Item> {...defaultProps} />
    )

    for (const column of defaultProps.columns) {
      expect(getByText(column)).toBeInTheDocument()

      expect(getByTestId(`column-counter-${column}`)).toHaveTextContent(
        (defaultProps.items[column as 'A'] ?? [])
          .filter(item => !isEmptyItemPlaceholder(item))
          .length.toString()
      )
    }
  })

  test('renders custom column label', () => {
    const { getByText } = render(
      <CardOrganizer<(typeof defaultProps.columns)[number], Item>
        {...defaultProps}
        renderColumnLabel={column => <span>{column} label</span>}
      />
    )

    for (const column of defaultProps.columns) {
      expect(getByText(column + ' label')).toBeInTheDocument()
    }
  })

  test('renders every card in the correct column', () => {
    const { getByRole, getByTestId } = render(
      <CardOrganizer<(typeof defaultProps.columns)[number], Item> {...defaultProps} />
    )

    for (const column of defaultProps.columns) {
      expect(getByTestId(`column-${column}`)).toBeInTheDocument()

      for (const item of defaultProps.items[column as 'A'] ?? []) {
        if (isEmptyItemPlaceholder(item)) {
          continue
        }
        const itemElement = getByRole('button', { name: item.name })
        expect(getByTestId(`column-${column}`).contains(itemElement))
      }
    }
  })

  test('calls onChange with the updated items when moving Up', async () => {
    const itemName = defaultProps.items['A'][1].name

    const onChange = vi.fn().mockResolvedValue(new Promise(() => {}))

    const { getByRole, findByText, asFragment, findByRole } = render(
      <CardOrganizer<(typeof defaultProps.columns)[number], Item>
        {...defaultProps}
        onChange={onChange}
      />
    )

    // Wait for the item to be rendered
    expect(await findByRole('button', { name: itemName })).toBeInTheDocument()

    const card = getByRole('button', { name: itemName })

    // Select card
    act(() => {
      fireEvent.keyDown(card, { code: 'Space' })
    })

    // Move card up
    act(() => {
      fireEvent.keyDown(card, { code: 'ArrowUp' })
    })
    await findByText('Draggable item 2 was moved over droppable area 1.')

    // Commit the change
    act(() => {
      fireEvent.keyDown(card, { code: 'Space' })
    })
    await findByText('Draggable item 2 was dropped over droppable area 1')

    expect(asFragment()).toMatchSnapshot()

    expect(onChange).toHaveBeenNthCalledWith(1, {
      from: expect.objectContaining({
        id: 2,
        sortable: expect.objectContaining({ containerId: 'A' }),
      }),
      to: expect.objectContaining({
        id: 1,
        sortable: expect.objectContaining({ containerId: 'A' }),
      }),
    })
  })

  test('calls onChange with the updated items when moving Down', async () => {
    const itemName = defaultProps.items['A'][0].name

    const onChange = vi.fn().mockResolvedValue(new Promise(() => {}))

    const { getByRole, findByText, asFragment, findByRole } = render(
      <CardOrganizer<(typeof defaultProps.columns)[number], Item>
        {...defaultProps}
        onChange={onChange}
      />
    )

    // Wait for the item to be rendered
    expect(await findByRole('button', { name: itemName })).toBeInTheDocument()

    const card = getByRole('button', { name: itemName })

    // Select card
    act(() => {
      fireEvent.keyDown(card, { code: 'Space' })
    })

    // Move card down
    act(() => {
      fireEvent.keyDown(card, { code: 'ArrowDown' })
    })
    await findByText('Draggable item 1 was moved over droppable area 1.')

    // Commit the change
    act(() => {
      fireEvent.keyDown(card, { code: 'Space' })
    })
    await findByText('Draggable item 1 was dropped over droppable area 1')

    expect(asFragment()).toMatchSnapshot()

    expect(onChange).toHaveBeenNthCalledWith(1, {
      from: expect.objectContaining({
        id: 1,
        sortable: expect.objectContaining({ containerId: 'A' }),
      }),
      to: expect.objectContaining({
        id: 1,
        sortable: expect.objectContaining({ containerId: 'A' }),
      }),
    })
  })

  test('does not call onChange when there are no changes or mouvement is canceled', async () => {
    const itemName = defaultProps.items['A'][1].name

    const onChange = vi.fn().mockResolvedValue(new Promise(() => {}))

    const { getByRole, asFragment, findByText, findByRole } = render(
      <CardOrganizer<(typeof defaultProps.columns)[number], Item>
        {...defaultProps}
        onChange={onChange}
      />
    )

    // Wait for the item to be rendered
    expect(await findByRole('button', { name: itemName })).toBeInTheDocument()

    const card = getByRole('button', { name: itemName })

    // Triggers card selection
    act(() => {
      fireEvent.keyDown(card, { code: 'Space' })
    })

    // Move card up
    await act(() => {
      fireEvent.keyDown(card, { code: 'ArrowUp' })
    })
    await findByText('Draggable item 2 was moved over droppable area 1.')

    // Cancel the change
    act(() => {
      fireEvent.keyDown(card, { code: 'Escape' })
    })

    expect(asFragment()).toMatchSnapshot()

    expect(onChange).toHaveBeenCalledTimes(0)
  })
})
