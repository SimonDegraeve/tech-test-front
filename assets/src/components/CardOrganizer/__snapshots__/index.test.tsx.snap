// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardOrganizer > calls onChange with the updated items 1`] = `
<DocumentFragment>
  <div
    class="sc-bXCLTC gghOHk"
  >
    <div
      class="sc-bXCLTC epIQNw"
    >
      <div
        class="sc-bXCLTC ioIWcI"
      >
        <div
          class="sc-bXCLTC bNzrHL"
        >
          <p
            class="sc-dxcDKg kVbkpz  wui-text"
            color="black"
          >
            A
          </p>
          <div
            class="sc-gvZAcH btjirq"
            data-testid="column-counter-A"
          >
            2
          </div>
        </div>
        <div
          class="sc-bXCLTC bMHor"
        >
          <div
            aria-describedby="DndDescribedBy-3"
            aria-disabled="false"
            aria-roledescription="sortable"
            role="button"
            style=""
            tabindex="0"
          >
            Item 2
          </div>
          <div
            aria-describedby="DndDescribedBy-3"
            aria-disabled="false"
            aria-roledescription="sortable"
            role="button"
            style=""
            tabindex="0"
          >
            Item 1
          </div>
        </div>
      </div>
      <div
        class="sc-bXCLTC ioIWcI"
      >
        <div
          class="sc-bXCLTC bNzrHL"
        >
          <p
            class="sc-dxcDKg kVbkpz  wui-text"
            color="black"
          >
            B
          </p>
          <div
            class="sc-gvZAcH btjirq"
            data-testid="column-counter-B"
          >
            0
          </div>
        </div>
        <div
          class="sc-bXCLTC bMHor"
        >
          <div
            aria-describedby="DndDescribedBy-3"
            aria-disabled="false"
            aria-roledescription="sortable"
            role="button"
            style=""
            tabindex="0"
          />
        </div>
      </div>
      <div
        class="sc-bXCLTC ioIWcI"
      >
        <div
          class="sc-bXCLTC bNzrHL"
        >
          <p
            class="sc-dxcDKg kVbkpz  wui-text"
            color="black"
          >
            C
          </p>
          <div
            class="sc-gvZAcH btjirq"
            data-testid="column-counter-C"
          >
            0
          </div>
        </div>
        <div
          class="sc-bXCLTC bMHor"
        >
          <div
            aria-describedby="DndDescribedBy-3"
            aria-disabled="false"
            aria-roledescription="sortable"
            role="button"
            style=""
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    id="DndDescribedBy-3"
    style="display: none;"
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-3"
    role="status"
    style="position: fixed; top: 0px; left: 0px; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  >
    Draggable item 2 was dropped over droppable area 1
  </div>
</DocumentFragment>
`;

exports[`CardOrganizer > does not call onChange when there are no changes or mouvement is canceled 1`] = `
<DocumentFragment>
  <div
    class="sc-bXCLTC gghOHk"
  >
    <div
      class="sc-bXCLTC epIQNw"
    >
      <div
        style="position: fixed; width: 0px; height: 0px; top: 0px; left: 0px; transform: translate3d(0px, 0px, 0) scaleX(1) scaleY(1); transition: transform 250ms ease; z-index: 999;"
      >
        <div
          style="transform: rotate(3deg);"
        >
          Item 2
        </div>
      </div>
      <div
        class="sc-bXCLTC ioIWcI"
      >
        <div
          class="sc-bXCLTC bNzrHL"
        >
          <p
            class="sc-dxcDKg kVbkpz  wui-text"
            color="black"
          >
            A
          </p>
          <div
            class="sc-gvZAcH btjirq"
            data-testid="column-counter-A"
          >
            2
          </div>
        </div>
        <div
          class="sc-bXCLTC bMHor"
        >
          <div
            aria-describedby="DndDescribedBy-4"
            aria-disabled="false"
            aria-roledescription="sortable"
            role="button"
            style="transition: transform 200ms ease; transform: translate3d(0px, 0px, 0) scaleX(1) scaleY(1);"
            tabindex="0"
          >
            Item 1
          </div>
          <div
            aria-describedby="DndDescribedBy-4"
            aria-disabled="false"
            aria-pressed="true"
            aria-roledescription="sortable"
            role="button"
            style="transition: transform 200ms ease; opacity: 0.5; transform: translate3d(0px, 0px, 0) scaleX(1) scaleY(1);"
            tabindex="0"
          >
            Item 2
          </div>
        </div>
      </div>
      <div
        class="sc-bXCLTC ioIWcI"
      >
        <div
          class="sc-bXCLTC bNzrHL"
        >
          <p
            class="sc-dxcDKg kVbkpz  wui-text"
            color="black"
          >
            B
          </p>
          <div
            class="sc-gvZAcH btjirq"
            data-testid="column-counter-B"
          >
            0
          </div>
        </div>
        <div
          class="sc-bXCLTC bMHor"
        >
          <div
            aria-describedby="DndDescribedBy-4"
            aria-disabled="false"
            aria-roledescription="sortable"
            role="button"
            style="transition: transform 200ms ease;"
            tabindex="0"
          />
        </div>
      </div>
      <div
        class="sc-bXCLTC ioIWcI"
      >
        <div
          class="sc-bXCLTC bNzrHL"
        >
          <p
            class="sc-dxcDKg kVbkpz  wui-text"
            color="black"
          >
            C
          </p>
          <div
            class="sc-gvZAcH btjirq"
            data-testid="column-counter-C"
          >
            0
          </div>
        </div>
        <div
          class="sc-bXCLTC bMHor"
        >
          <div
            aria-describedby="DndDescribedBy-4"
            aria-disabled="false"
            aria-roledescription="sortable"
            role="button"
            style="transition: transform 200ms ease;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    id="DndDescribedBy-4"
    style="display: none;"
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-4"
    role="status"
    style="position: fixed; top: 0px; left: 0px; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  >
    Draggable item 2 was moved over droppable area 1.
  </div>
</DocumentFragment>
`;
